qadence/mitigations/readout.py

```diff
@@ -25,51 +57,76 @@ def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArr
+        mitigation: Selects additional mitigation options based on noise choice.
+        For readout we have the following mitigation options for optimization
+        1.constrained_opt, 2.mle. Default : mle
+        samples: List of samples to be mitigated
+
+    Returns:
+        Mitigated counts computed by the algorithm
     """
     noise_matrices = noise.options.get("noise_matrix", noise.options["confusion_matrices"])
+    optimization_type = mitigation.options.get("optimization_type", "mle")
     n_qubits = len(list(samples[0].keys())[0])
     n_shots = sum(samples[0].values())
     # Build the whole T matrix.
     T_matrix = reduce(torch.kron, noise_matrices).detach().numpy()
     corrected_counters: list[Counter] = []
+
+    if optimization_type == "mle":  ## MLE version
+        ##check if matrix is singular and use appropriate inverse
+        T_inv = inv(T_matrix) if matrix_rank(T_matrix) == T_matrix.shape[0] else pinv(T_matrix)
+
     for sample in samples:
         bitstring_length = 2**n_qubits
         # List of bitstrings in lexicographical order.
         ordered_bitstrings = [f"{i:0{n_qubits}b}" for i in range(bitstring_length)]
         # Array of raw probabilites.
         p_raw = np.array([sample[bs] for bs in ordered_bitstrings]) / n_shots
-        # Initial random guess in [0,1].
-        p_corr0 = np.random.rand(bitstring_length)
-        # Stochasticity constraints.
-        normality_constraint = LinearConstraint(
-            np.ones(bitstring_length).astype(int), lb=1.0, ub=1.0
-        )
-        positivity_constraint = LinearConstraint(
-            np.eye(bitstring_length).astype(int), lb=0.0, ub=1.0
-        )
-        constraints = [normality_constraint, positivity_constraint]
-        # Minimize the corrected probabilities.
-        res = minimize(corrected_probas, p_corr0, args=(T_matrix, p_raw), constraints=constraints)
-        # breakpoint()
-        corrected_counts = np.rint(res.x * n_shots).astype(int)
+
+        if optimization_type == "constrained_opt":
+            # Initial random guess in [0,1].
+            p_corr0 = np.random.rand(bitstring_length)
+            # Stochasticity constraints.
+            normality_constraint = LinearConstraint(
+                np.ones(bitstring_length).astype(int), lb=1.0, ub=1.0
+            )
+            positivity_constraint = LinearConstraint(
+                np.eye(bitstring_length).astype(int), lb=0.0, ub=1.0
+            )
+            constraints = [normality_constraint, positivity_constraint]
+            # Minimize the corrected probabilities.
+            res = minimize(
+                corrected_probas, p_corr0, args=(T_matrix, p_raw), constraints=constraints
+            )
+            p_corr = res.x
+
+        elif optimization_type == "mle":  ## MLE version
+            ## compute corrected inverse using matrix inversion and run MLE
+            p_corr = mle_solve(T_inv @ p_raw)
```

Comment:
OK thanks. But this is not what you do here as far as I can see. You still invert T_matrix either exactly or pseudo depending on its singularity ?

Reply:
Since the noise is coming in as independent T matrices, I think I will invert them independently and multiply. Yes I would be applying inverse or pseduo inverse on 2x2 matrices based on the singularity captured by rank